addons/content-index/src/main/java/org/commonjava/indy/content/index/conf/HostedContentIndexRescanManager.java

```diff
@@ -0,0 +1,120 @@
+import org.commonjava.indy.data.StoreDataManager;
+import org.commonjava.indy.model.core.ArtifactStore;
+import org.commonjava.indy.model.core.Group;
+import org.commonjava.indy.model.core.HostedRepository;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.model.core.StoreType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This component will handle the content index for the hosted repo during its rescan. It will rebuild all relevant
+ * index entries for the hosted repo, and remove all useless entries for the missed artifacts in hosted repo
+ */
+@ApplicationScoped
+public class HostedContentIndexRescanManager
+{
+    private final Logger LOGGER = LoggerFactory.getLogger( this.getClass() );
+
+    @Inject
+    private StoreDataManager storeDataManager;
+
+    @Inject
+    private ContentIndexManager contentIndexManager;
+
+    @Inject
+    private ContentManager contentManager;
+
+    protected HostedContentIndexRescanManager()
+    {
+    }
+
+    public void hostedIndexPreRescan( @Observes final ArtifactStorePreRescanEvent e )
+            throws IndyWorkflowException
+    {
+        Collection<ArtifactStore> affectedRepos = e.getStores();
+        for ( ArtifactStore repo : affectedRepos )
+        {
+            if ( repo.getType() == StoreType.hosted )
+            {
+                final HostedRepository hosted = (HostedRepository) repo;
+
+                LOGGER.trace( "Clear content index for {}", hosted.getKey() );
+                // Remove the content index items for the hosted which will be rescanned
+                contentIndexManager.clearAllIndexedPathInStore( hosted );
+                // Remove the content index items for the affected groups of the hosted which will be rescanned, note that
+                // we will only cared about the items that is from this hosted only but not others(the origin key in
+                // IndexedStorePath which hits this hosted)
+                contentIndexManager.clearAllIndexedPathWithOriginalStore( hosted );
+            }
+        }
+    }
+
+    public void hostedIndexPostRescan( @Observes final ArtifactStorePostRescanEvent e )
+            throws IndyWorkflowException
+    {
+        Collection<ArtifactStore> hostedStores = e.getStores();
+        for ( ArtifactStore repo : hostedStores )
+        {
+            if ( repo.getType() == StoreType.hosted )
+            {
+                LOGGER.trace( "Rebuild content index for {}", repo.getKey() );
+                final HostedRepository hosted = (HostedRepository) repo;
+                try
+                {
+                    List<StoreResource> resources = contentManager.list( hosted, DownloadManager.ROOT_PATH );
```

Comment:
This won't list recursively, I don't think. If we have content under the path `org/commonjava/commonjava/11/commonjava-11.pom` and get a listing for the root, I don't think it'll be included in the result.